src/psyclone/psyir/nodes/intrinsic_call.py

```diff
@@ -90,655 +422,2096 @@ class IntrinsicCall(Call):
-            ArgDesc(0, 0, None), {})
+            "COMMAND_ARGUMENT_COUNT",
+            True,
+            False,
+            False,
+            ArgDesc(0, 0, None),
+            {},
+            INTEGER_TYPE,
+            None,
+        )
         CONJG = IAttr(
-            'CONJG', True, True, False,
-            ArgDesc(1, 1, DataNode), {})
+            "CONJG",
+            True,
+            True,
+            False,
+            ArgDesc(1, 1, DataNode),
+            {},
+            # TODO #1590 Complex numbers unsupported.
+            lambda node: UnresolvedType(),
+            None,
+        )
         COS = IAttr(
-            'COS', True, True, False,
-            ArgDesc(1, 1, DataNode), {})
+            "COS",
+            True,
+            True,
+            False,
+            ArgDesc(1, 1, DataNode),
+            {},
+            _get_first_argument_type,
+            None,
+        )
         COSH = IAttr(
-            'COSH', True, True, False,
-            ArgDesc(1, 1, DataNode), {})
+            "COSH",
+            True,
+            True,
+            False,
+            ArgDesc(1, 1, DataNode),
+            {},
+            _get_first_argument_type,
+            None,
+        )
         COSHAPE = IAttr(
-            'COSHAPE', True, False, True,
-            ArgDesc(1, 1, DataNode), {"kind": DataNode})
+            "COSHAPE",
+            True,
+            False,
+            True,
+            ArgDesc(1, 1, DataNode),
+            {"kind": DataNode},
+            # FIXME Return type
+            lambda node: ArrayType(
+                ScalarType(
+                    ScalarType.Intrinsic.INTEGER,
+                    (
+                        node.arguments[0].datatype.precision
+                        if "kind" not in node.argument_names
+                        else node.arguments[node.argument_names.index("kind")]
+                    ),
+                ),
+                [
+                    (
+                        index.copy()
+                        if not isinstance(index, ArrayType.ArrayBounds)
+                        else ArrayType.ArrayBounds(index.lower, index.upper)
+                    )
+                    for index in node.arguments[0].datatype.shape
+                ],
+            ),
+            None,
+        )
         COUNT = IAttr(
-            'COUNT', True, False, False,
-            ArgDesc(1, 1, DataNode), {"dim": DataNode, "kind": DataNode})
+            "COUNT",
+            True,
+            False,
+            False,
+            ArgDesc(1, 1, DataNode),
+            {"dim": DataNode, "kind": DataNode},
+            _get_integer_of_kind_with_optional_dim,
+            None,
+        )
         CPU_TIME = IAttr(
-            'CPU_TIME', False, False, False,
-            ArgDesc(1, 1, DataNode), {})
+            "CPU_TIME",
+            False,
+            False,
+            False,
+            ArgDesc(1, 1, DataNode),
+            {},
+            None,
```

Comment:
Strictly speaking, as this is a call to a routine, the type is `lambda: NoType()`. I was going to say we need to make sure this is consistent with `Call.datatype` but that method isn't implemented yet :-)